.github/workflows/template.flink-ci.yml

```diff
@@ -0,0 +1,438 @@
+  DOCKER_IMAGES_CACHE_FOLDER: /root/.docker-cache
+
+jobs:
+  compile:
+    name: "Compile"
+    runs-on: ubuntu-latest
+    container:
+      image: chesnay/flink-ci:java_8_11_17_21_maven_386
+      options: --init
+    timeout-minutes: 240
+    outputs:
+      stringified-workflow-name: ${{ steps.workflow-prep-step.outputs.stringified-workflow-name }}
+    steps:
+      - name: "Flink Checkout"
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ inputs.branch }}
+          persist-credentials: false
+
+      - name: "Stringify workflow name"
+        id: workflow-prep-step
+        run: |
+          stringified_workflow_name=$(echo "${{ github.workflow }}-${{ inputs.workflow-caller-id }}" | tr -C '[:alnum:]._' '-' |  tr '[:upper:]' '[:lower:]' | sed -e 's/--*/-/g' -e 's/^-*//g' -e 's/-*$//g')
+          echo "stringified-workflow-name=${stringified_workflow_name}" >> $GITHUB_OUTPUT
+
+      - name: "Set JDK version to Java ${{ inputs.jdk_version }}"
+        uses: "./.github/actions/set_java_in_container"
+        with:
+          jdk_version: ${{ inputs.jdk_version }}
+
+      - name: "Setup Maven package cache"
+        uses: actions/cache@v3
+        with:
+          path: ${{ env.MAVEN_REPO_FOLDER }}
+          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+          restore-keys: ${{ runner.os }}-maven-
+
+      - name: "Moves relevant checkout content to container-local working directory"
+        uses: "./.github/actions/move_checkout"
+        with:
+          source_directory: ${{ env.MOUNTED_WORKING_DIR }}
+          target_directory: ${{ env.CONTAINER_LOCAL_WORKING_DIR }}
+
+      - name: "Compile Flink"
+        uses: "./.github/actions/run_mvn"
+        with:
+          working_directory: ${{ env.CONTAINER_LOCAL_WORKING_DIR }}
+          maven-parameters: "test-compile -Dflink.markBundledAsOptional=false -Dfast"
+          env: "${{ inputs.environment }}"
+
+      - name: "Collect build artifacts"
+        working-directory: ${{ env.CONTAINER_LOCAL_WORKING_DIR }}
+        run: |
+          ./tools/azure-pipelines/create_build_artifact.sh
+
+      - name: "Upload artifacts to make them available in downstream jobs"
+        uses: actions/upload-artifact@v3
+        with:
+          name: build-artifacts-${{ steps.workflow-prep-step.outputs.stringified-workflow-name }}-${{ github.run_number }}
+          path: ${{ env.FLINK_ARTIFACT_DIR }}/${{ env.FLINK_ARTIFACT_FILENAME }}
+          if-no-files-found: error
+          # use minimum here because we only need these artifacts to speed up the build
+          retention-days: 1
+
+  packaging:
+    name: "Test packaging/licensing"
+    needs: compile
+    runs-on: ubuntu-latest
+    container:
+      image: chesnay/flink-ci:java_8_11_17_21_maven_386
+      options: --init
+
+    steps:
+      - name: "Flink Checkout"
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ inputs.branch }}
+          persist-credentials: false
```

Comment:
I was wondering the same. But decided against it because the lines of code would not reduce much. Only the `persist-credentials` line would be removed. I thought that this wouldn't be worth an extra artifact.

Reply:
I revisited this one and created a `job_init` composite action. But I realized then that we cannot add the checkout action: The local composite actions need to be checked out before using them. An alternative could be to have the composite actions being deployed somewhere else through a dedicated "flink-ci" branch or repo. But I'd consider that to be a follow-up.